### PR TITLE
Get new RelativeFilePaths if previous do not exist

### PR DESCRIPTION
**Pull request by mdanish-kh**
- [x] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- [x] Are you working against an Issue?
  - #414

This change makes it so we take the last portion of the RelativeFilePath i.e., the file name and search that file name in the new archive for a match. If it is a unique match, we get its RelativeFilePath and use it for the update.

Known limitation is that there are some cases where there may be more than one match in the archive (An example of this being [DynamoRIO.DynamoRIO package](https://github.com/microsoft/winget-pkgs/blob/master/manifests/d/DynamoRIO/DynamoRIO/9.93.19518/DynamoRIO.DynamoRIO.installer.yaml)). This is not handled in this PR and we print out an appropriate message to the user. I'll create an issue for it if this PR gets accepted.

The change to use the whole NestedInstallerFile object instead of previous RelativeFilePaths is so that I can replace this whole object in UpdateInstallerMetadata() and have it updated.

### Test packages

Somes packages I found to test this PR with

```powershell
wingetcreate update Smallstep.step -v 18.0 --urls "https://github.com/smallstep/cli/releases/download/v0.24.4/step_windows_0.24.4_amd64.zip" "https://github.com/smallstep/cli/releases/download/v0.24.4/step_windows_0.24.4_arm64.zip"
```

```powershell
wingetcreate update Genymobile.scrcpy -v 18.0 --urls "https://github.com/Genymobile/scrcpy/releases/download/v1.25/scrcpy-win64-v1.25.zip" "https://github.com/Genymobile/scrcpy/releases/download/v1.25/scrcpy-win32-v1.25.zip"
```

Multiple matches for RelativeFilePaths (**not handled in this PR**)
```powershell
wingetcreate update DynamoRIO.DynamoRIO --version 18.0 --urls "https://github.com/DynamoRIO/dynamorio/releases/download/cronbuild-9.93.19557/DynamoRIO-Windows-9.93.19557.zip"
```

-----

 ###### Microsoft Reviewers: codeflow:open?pullrequest=https://github.com/microsoft/winget-create/pull/415&drop=dogfoodAlpha

### CODE REVIEW

**Comment by mdanish-kh** [comment]
The GitHub diffs are not correct after https://github.com/microsoft/winget-create/pull/415/commits/83973aa632218217c09ea7a330898a3a8e0083c0. I only modified the UpdateSingleInstallerInteractively method but that method had both mixed line endings (CRLF and LF) because of which the diffs are messed up. If you check out the changes in Visual Studio, it'll show you the correct diffs. 
- Related to https://github.com/microsoft/winget-create/issues/391

File: src/WingetCreateCLI/Commands/UpdateCommand.cs [hunk 1]
```diff
@@ -629,124 +673,140 @@ private List<InstallerMetadata> ParseInstallerUrlsForOverrides(List<string> inst
...
-                    else
-                    {
-                        Logger.ErrorLocalized(nameof(Resources.UnmatchedNestedInstaller_Error), relativeFilePath, url);
-                        continue;
+
+                        // Update the relative file path for all nested installers.
+                        foreach (NestedInstallerFile nestedInstaller in installer.NestedInstallerFiles)
+                        {
+                            nestedInstaller.RelativeFilePath = ObtainMatchingRelativeFilePath(nestedInstaller.RelativeFilePath, extractDirectory, packageFile);
+                            if (nestedInstaller.RelativeFilePath == null)
```
Comment (ryfu-msft) [inline src/WingetCreateCLI/Commands/UpdateCommand.cs:805]: nit:
```suggestion
                            if (string.IsNullOrEmpty(nestedInstaller.RelativeFilePath))

```

File: src/WingetCreateCLI/Commands/UpdateCommand.cs [hunk 2]
```diff
@@ -419,6 +440,29 @@ private static string ExtractArchiveAndRetrieveDirectoryPath(string packageFileP
...
 
+        private static string ObtainMatchingRelativeFilePath(string oldRelativeFilePath, string directory, string archiveName)
+        {
+            string fileName = Path.GetFileName(oldRelativeFilePath);
+            string[] matchingFiles = Directory.GetFiles(directory, fileName, SearchOption.AllDirectories);
+
+            // If there is only one match in the archive, use that as the new relative file path.
+            if (matchingFiles.Length == 1)
+            {
+                string relativePath = matchingFiles.First().Replace(directory, string.Empty).TrimStart('\\');
```
Comment (ryfu-msft) [inline src/WingetCreateCLI/Commands/UpdateCommand.cs:451]:
```suggestion
                string relativePath = matchingFiles.First().Replace(directory, string.Empty).TrimStart(Path.DirectorySeparatorChar);

```

File: src/WingetCreateCLI/Commands/UpdateCommand.cs [hunk 3]
```diff
@@ -280,21 +280,42 @@ public async Task<Manifests> UpdateManifestsAutonomously(Manifests manifests)
...
                     {
-                        if (!File.Exists(Path.Combine(extractDirectory, relativeFilePath)))
+                        // If the previous RelativeFilePath is not an exact match, check if there is a new suitable match in the archive.
+                        if (!File.Exists(Path.Combine(extractDirectory, nestedInstallerFile.RelativeFilePath)))
                         {
-                            Logger.ErrorLocalized(nameof(Resources.NestedInstallerFileNotFound_Error), relativeFilePath);
-                            return null;
-                        }
+                            string relativeFilePath = ObtainMatchingRelativeFilePath(nestedInstallerFile.RelativeFilePath, extractDirectory, packageFile);
+                            if (relativeFilePath == null)
```
Comment (ryfu-msft) [inline src/WingetCreateCLI/Commands/UpdateCommand.cs:294]: nit:
```suggestion
                if (string.IsNullOrEmpty(packageFile))
```

File: src/WingetCreateCLI/Commands/UpdateCommand.cs [hunk 4]
```diff
@@ -280,21 +280,42 @@ public async Task<Manifests> UpdateManifestsAutonomously(Manifests manifests)
...
+                            {
+                                return null;
+                            }
+
+                            List<string> relativeFilePathsInsideList = installerUpdate.NestedInstallerFiles.Select(x => x.RelativeFilePath).ToList();
```
Comment (ryfu-msft) [inline src/WingetCreateCLI/Commands/UpdateCommand.cs:299]: Nit: This variable name was slightly confusing for me. I think something like `installerUpdateRelativeFilePathList` might be better.

File: src/WingetCreateCLI/Commands/UpdateCommand.cs [hunk 5]
```diff
@@ -280,21 +280,42 @@ public async Task<Manifests> UpdateManifestsAutonomously(Manifests manifests)
...
 
-                        installerUpdate.RelativeFilePaths.Add(relativeFilePath);
+                            // Skip if the relative path is already in the list.
```
Comment (ryfu-msft) [inline src/WingetCreateCLI/Commands/UpdateCommand.cs:301]:
```suggestion
                            // Skip if the relative path is already in the installer update list.

```

**Review by ryfu-msft — CHANGES_REQUESTED** [review]
Just a couple nit comments

**Comment by ryfu-msft** [comment]
> The GitHub diffs are not correct after [83973aa](https://github.com/microsoft/winget-create/commit/83973aa632218217c09ea7a330898a3a8e0083c0). I modified the UpdateSingleInstallerInteractively method but that method had both mixed line endings (CRLF and LF) because of which the diffs are messed up. If you check out the changes in Visual Studio, it'll show you the correct diffs.
> 
> * Related to [CRLF line endings for all sources files #391](https://github.com/microsoft/winget-create/issues/391)
> 
> If you like I can fix the line endings for UpdateCommand.cs and PackageParser.cs in a new commit as they are the two most high traffic files, though that'll mess the diffs up even more for this PR

If you can do this in a separate PR (after this gets checked in) that would be a huge help! 🙏 
That way we can link the related issue to that PR when it gets checked in.

**Comment by mdanish-kh** [comment]
> If you can do this in a separate PR (after this gets checked in) that would be a huge help! 🙏 That way we can link the related issue to that PR when it gets checked in.

Gotcha! Will do after this one. I was shying away from fixing the line endings since it'll mess up the `git blame` output but I guess fixing it has more value.

**Review by ryfu-msft — APPROVED** [review]

**Pull request merged**